_posts/2022-04-13-ambient-peer-discovery.md

```diff
@@ -0,0 +1,254 @@
+If Waku v2 would do the same, only a small subset of the retrieved nodes would support Waku v2.
+
+A first naive solution for Waku v2 discv5 discovery is
+
+* query for a random node set
+* filter the returned nodes on the query path based on Waku v2 capability via the [Waku v2 ENR](https://rfc.vac.dev/spec/31/)
+* repeat until enough Waku v2 capable nodes are found
+
+This query process boils down to random walk discovery, which is very resilient against attacks, but also very inefficient if the number of nodes supporting the desired capability is small.
+We refer to this is as the needle-in-the-haystack problem.
+
+### Random Walk Performance Estimation
+
+This subsection provides a rough estimation on the overhead introduced by random walk discovery.
+
+Given the following parameters:
+
+* $n$ number of total nodes participating in discv5
+* $p$ percentage of nodes supporting Waku
+* $W$ the event of having at least one Waku node in a random sample
+* $k$ the size of a random sample (default = 16)
+* $\alpha$ the number of parallel queries started
+* $b$ bits per hop
+* $q$ the number of queries
+
+A query takes $log_{2^b}n$ hops to retrieve a random sample of nodes.
+
+$P(W) = 1 - (1-p/100)^k$ is the probability of having at least one Waku node in the sample.
+
+$P(W^q) = 1 - (1-p/100)^{kq}$  is the probability of having at least one Waku node in the union of $q$ samples.
+
+Expressing this in terms of $q$, we can write:
+$$P(W^q) = 1 - (1-p/100)^{kq} \iff  q = log_{(1-p/100)^k}(1-P(W^q))$$
+
+Figure 1 shows a log-log plot for $P(W^q) = 90\%$.
+
+<p align="center">
+    <img src="../assets/img/waku_v2_discv5_random_walk_estimation.svg"  width="50%" />
+    <br />
+    Figure 1: log-log plot showing the number of queries necessary to retrieve a Waku v2 node with a probability of 90% in relation to the precentage of Waku v2 nodes in the network.
+</p>
+
+Assuming $p=0.1$, we would need
+
+$$0.9 = 1 - (1-0.1/100)^{16q} => q \approx 144$$
+
+queries to get a Waku node with 90% probability, which leads to $\approx 144 * 18 = 2592$ overlay hops.
+Choosing $b=3$ would reduce the number to $\approx 144 * 6 = 864$.
+Even when choosing $\alpha = 10$ we would have to wait at least 80 RTTs.
+This effort is just for retrieving a single Waku node. Ideally, we want at least 3 Waku nodes for bootstrapping a Waku relay.
+
+[The discv5 docu](https://github.com/ethereum/devp2p/blob/6b0abc3d956a626c28dce1307ee9f546db17b6bd/discv5/discv5-theory.md#ad-placement-and-topic-radius) roughly estimates $p=1%$ to be the threshold for acceptably efficient random walk discovery. 
+This is in line with our estimation:
+
+$$0.9 = 1 - (1-1/100)^{16q} => q \approx 14$$
+
+
+The number of necessary queries is linearly dependent on the percentage $p$ of Waku nodes.
+The number of hops per query is logarithmically dependent on $n$.
+Thus, random walk searching is inefficient for small percentages $p$.
+However, random walks are more resilient against attacks.
+
```

Comment:
Comment above belongs here

This is the nice segue into "we need something better"